wiring/src/spark_wiring_ble.cpp

```diff
@@ -577,81 +576,81 @@ BleAdvertisingData::BleAdvertisingData(const iBeacon& beacon)
     // Length of the following payload
-    selfData_[selfLen_++] = 0x15;
+    selfData_.append(0x15);
     // 128-bits Beacon UUID, MSB
     for (size_t i = 0, j = BLE_SIG_UUID_128BIT_LEN - 1; i < BLE_SIG_UUID_128BIT_LEN; i++, j--) {
-        selfData_[selfLen_ + i] = beacon.UUID().rawBytes()[j];
+        selfData_.append(beacon.UUID().rawBytes()[j]);
     }
-    selfLen_ += BLE_SIG_UUID_128BIT_LEN;
     // Major, MSB
-    selfData_[selfLen_++] = (uint8_t)((beacon.major() >> 8) & 0x00FF);
-    selfData_[selfLen_++] = (uint8_t)(beacon.major() & 0x00FF);
+    selfData_.append((uint8_t)((beacon.major() >> 8) & 0x00FF));
+    selfData_.append((uint8_t)(beacon.major() & 0x00FF));
     // Minor, MSB
-    selfData_[selfLen_++] = (uint8_t)((beacon.minor() >> 8) & 0x00FF);
-    selfData_[selfLen_++] = (uint8_t)(beacon.minor() & 0x00FF);
+    selfData_.append((uint8_t)((beacon.minor() >> 8) & 0x00FF));
+    selfData_.append((uint8_t)(beacon.minor() & 0x00FF));
     // Measure power
-    selfData_[selfLen_++] = beacon.measurePower();
+    selfData_.append(beacon.measurePower());
+    selfData_.trimToSize();
 
-    return selfLen_;
+    return selfData_.size();
 }
 
 size_t BleAdvertisingData::append(BleAdvertisingDataType type, const uint8_t* buf, size_t len, bool force) {
     if (buf == nullptr) {
-        return selfLen_;
+        return selfData_.size();
     }
     size_t offset;
-    size_t adsLen = locate(selfData_, selfLen_, type, &offset);
+    size_t adsLen = locate(selfData_.data(), selfData_.size(), type, &offset);
     if (!force && adsLen > 0) {
         // Update the existing AD structure.
-        size_t staLen = selfLen_ - adsLen;
-        if ((staLen + len + 2) <= BLE_MAX_ADV_DATA_LEN) {
-            // Firstly, move the last consistent block.
-            uint16_t moveLen = selfLen_ - offset - adsLen;
-            memmove(&selfData_[offset + len + 2], &selfData_[offset + adsLen], moveLen);
+        if ((selfData_.size() - adsLen + len + 2) <= BLE_MAX_ADV_DATA_LEN_EXT) {
+            // Firstly, remove the existing AD structure.
+            selfData_.removeAt(offset, adsLen);
             // Secondly, Update the AD structure.
+            // Reserve the RAM in the vector, and return if it fails
+            CHECK_TRUE(selfData_.reserve(selfData_.size() + len + 2), selfData_.size());
             // The Length field is the total length of Type field and Data field.
-            selfData_[offset] = len + 1;
-            memcpy(&selfData_[offset + 2], buf, len);
+            selfData_.insert(offset, len + 1);
             // An AD structure is composed of one byte length field, one byte Type field and Data field.
-            selfLen_ = staLen + len + 2;
+            selfData_.insert(offset + 1, static_cast<uint8_t>(type));
+            selfData_.insert(offset + 2, buf, len);
         }
     }
-    else if ((selfLen_ + len + 2) <= BLE_MAX_ADV_DATA_LEN) {
+    else if ((selfData_.size() + len + 2) <= BLE_MAX_ADV_DATA_LEN_EXT) {
```

Comment:
We should use `BLE_MAX_ADV_DATA_LEN_EXT_CONNECTABLE` here.

Reply:
See above.